### ISSUE TITLE
Minion did not return. [No response]

### ISSUE THREAD

**Issue by alangarf** [saltstack/salt#33784]
Hi,

We've upgraded recently to 2015.8.10, and now we're forever seeing:
```
    Minion did not return. [No response]
```
from our minions. This is infuriating as it gives no information nor does the master have any logs. We can see that _sometimes_ the minion will be doing the work and _sometimes_ a call to `salt-run jobs.active` will show the job running, but other times it just disappears into limbo.

We've also seen a call from a master not being sent to the minion and received a `Minion did not return. [No response]` but then subsequently sent the job again and then the minion gets *two* jobs and the duplicate job_id error.

Also a call to `salt-run jobs.active` will block and timeout or return no information.

---

I've removed the minion keys and cleared out all the keys, and that will fix it for a while *sometimes*. I've tried batching, I've tried targeting a single minion. It's just broken. Salt-call from the minions to do highstates work fine, but I don't want to use a "puppet-like" system, I want to do this work from the master!

---

I'm seeing this across the event bus:

```
20160606070300982476	{
    "_stamp": "2016-06-06T07:03:00.982692",
    "minions": [
        "prod-apia6.foo.com",
        "prod-apia8.foo.com"
    ]
}
salt/job/20160606070300982476/new	{
    "_stamp": "2016-06-06T07:03:00.982836",
    "arg": [],
    "fun": "test.ping",
    "jid": "20160606070300982476",
    "minions": [
        "prod-apia6.foo.com",
        "prod-apia8.foo.com"
    ],
    "tgt": "prod-apia[5678]*",
    "tgt_type": "glob",
    "user": "sudo_alan"
}

20160606070330994590	{
    "_stamp": "2016-06-06T07:03:30.994810",
    "minions": [
        "prod-apia6.foo.com",
        "prod-apia8.foo.com"
    ]
}
salt/job/20160606070330994590/new	{
    "_stamp": "2016-06-06T07:03:30.994969",
    "arg": [
        "20160606070300982476"
    ],
    "fun": "saltutil.find_job",
    "jid": "20160606070330994590",
    "minions": [
        "prod-apia6.foo.com",
        "prod-apia8.foo.com"
    ],
    "tgt": "prod-apia[5678]*",
    "tgt_type": "glob",
    "user": "sudo_alan"
}
```

And then I get:
```
$ sudo salt 'prod-apia[5678]*' test.ping
prod-apia8.va.lninfra.net:
    Minion did not return. [No response]
prod-apia6.va.lninfra.net:
    Minion did not return. [No response]
```

---

```
Salt Version:
           Salt: 2015.8.10

Dependency Versions:
         Jinja2: 2.7.2
       M2Crypto: Not Installed
           Mako: 0.9.1
         PyYAML: 3.10
          PyZMQ: 14.0.1
         Python: 2.7.6 (default, Jun 22 2015, 17:58:13)
           RAET: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.4
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: 1.5
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
        libgit2: Not Installed
        libnacl: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.3.0
   mysql-python: 1.2.3
      pycparser: Not Installed
       pycrypto: 2.6.1
         pygit2: Not Installed
   python-gnupg: Not Installed
          smmap: 0.8.2
        timelib: Not Installed

System Versions:
           dist: Ubuntu 14.04 trusty
        machine: x86_64
        release: 3.13.0-79-generic
         system: Ubuntu 14.04 trusty
```

**Comment by alangarf** [comment]
Also seeing errors like this occasionally, no idea why it's complaining about multi-master. We have only one on this stack.
```
TypeError: string indices must be integers, not str
2016-06-06 07:18:38,395 [salt.minion      ][ERROR   ][1788] Pillar data could not be refreshed. One or more masters may be down!
2016-06-06 07:18:38,734 [salt.pillar      ][ERROR   ][1788] Exception getting pillar:
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 121, in compile_pillar
    dictkey='pillar',
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 870, in run
    value = future.result()
  File "/usr/lib/python2.7/dist-packages/tornado/concurrent.py", line 215, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/dist-packages/tornado/gen.py", line 879, in run
    yielded = self.gen.send(value)
  File "/usr/lib/python2.7/dist-packages/salt/transport/zeromq.py", line 161, in crypted_transfer_decode_dictentry
    aes = cipher.decrypt(ret['key'])
TypeError: string indices must be integers, not str
```

**Comment by cachedout** [comment]
Did you just upgrade the minions or did you upgrade the master as well? The master should always be upgraded first.

**Comment by alangarf** [comment]
Yes we upgraded all the instances, they are all running 2015.8.10. We've even shut all the salt minion's and master down, removed all the pki dirs and started the master and then all the minions so they would re-register. This helped for a little while, but then it has all gone to bad again.

We're starting to see an issue where the "find_jobs" call watching the event bus with `salt-run state.event pretty=True` will take a *long* time to return. We've even tried adjusting the time out, but it looks like the minion is running jobs synchronously so it will not return a `test.ping` for example if another job is running (including a find_job). It's very weird. 

eg. a test.ping will run, the minion will get the job, it will pause for some reason (unknown at this stage), a "find_job" event will be raised targeting the minion, the "No response" error will pop out from the salt call, then on the event bus there will be the test.ping event followed start away by the "find_jobs" event response.

**Comment by cachedout** [comment]
Wow, yeah. This isn't something I've seen nor, frankly, even heard of happening. What do you think about maybe throwing some `strace` on that minion process and see if you can spot it hanging? Bear in mind you'll want to run that with a `-f` to follow forks as they happen.

**Comment by bemeyert** [comment]
It just now happened again. From approx. 160 minions 152 are just gone! A restart of the minion helps. As straces go, here from a minions that's according to the master down.. The first is from the parent process the second from it's fork.
```
# ps -ef | grep salt
root      1433     1  0 Jul07 ?        00:00:00 /usr/bin/python /usr/bin/salt-minion
root      3070  1433  0 Jul07 ?        00:10:44 /usr/bin/python /usr/bin/salt-minion
root     28283  5620  0 13:08 pts/1    00:00:00 grep --color=auto salt
# strace -f -p 1433
Process 1433 attached
wait4(3070, 
# strace -f -p 3070
Process 3070 attached with 6 threads
[pid  6487] epoll_wait(22,  <unfinished ...>
[pid  6429] epoll_wait(14,  <unfinished ...>
[pid  6488] epoll_wait(24,  <unfinished ...>
[pid  6428] epoll_wait(12,  <unfinished ...>
[pid  3075] select(0, NULL, NULL, NULL, {0, 850711} <unfinished ...>
[pid  3070] restart_syscall(<... resuming interrupted call ...> <unfinished ...>
[pid  3075] <... select resumed> )      = 0 (Timeout)
[pid  3075] getuid()                    = 0
[pid  3075] kill(1433, SIG_0)           = 0
[pid  3075] select(0, NULL, NULL, NULL, {5, 0} <unfinished ...>
[pid  3070] <... restart_syscall resumed> ) = 0
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 0) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 998) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 0) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 997) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 0) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 997) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 0) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 997) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 0) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 997 <unfinished ...>
[pid  3075] <... select resumed> )      = 0 (Timeout)
[pid  3075] getuid()                    = 0
[pid  3075] kill(1433, SIG_0)           = 0
[pid  3075] select(0, NULL, NULL, NULL, {5, 0} <unfinished ...>
[pid  3070] <... poll resumed> )        = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 0) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 997) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 0) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 997) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 0) = 0 (Timeout)
[pid  3070] poll([{fd=16, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=25, events=POLLIN}], 1, 0) = 0 (Timeout)
[pid  3070] poll([{fd=8, events=POLLIN}, {fd=16, events=POLLIN}, {fd=25, events=POLLIN}], 3, 997^CProcess 3070 detached
 <detached ...>
Process 3075 detached
Process 6428 detached
Process 6429 detached
Process 6487 detached
Process 6488 detached
```

And nothing in the logs!? This happens randomly and I don't know how to reproduce it.

**Comment by bemeyert** [comment]
Oh well. Now all minions are back after about 12 minutes (???). Only thing I did was 
- restart a single minion
- every minute a `manage.status` is running via cron
So it is possible that the above trace is worthless since while taking it the minion came back to live.

**Comment by bemeyert** [comment]
BTW: Same version (2015.8.10)

**Comment by bemeyert** [comment]
We still haven't ruled out our network as a culprit, although chances are low. "Hopefully" the next "outage" comes soon, so I can gather more information.

**Comment by bemeyert** [comment]
At least show the above strace the correct information, when the Minions are gone. Funny thing is that a `test.ping` from the master fails, but a `salt-call state.show_top` on the same minion connects cleanly to the master (verified via tcpdump) and works.
But we are still looking into the network...
Sry, if this comes in telegram style. But we can only debug, when the issue appears and the minions vanished. Which happened since yesterday too often. My gut tells me that this is a network issue. I will get back I know more.

**Comment by bemeyert** [comment]
OK, we don't have this issue. It was the fault of our network. We use OpenBSDs "pf" firewall. There is  a limit for the number of states called `adaptive.start` which we crossed with some new services having lots of persistent TCP connections. Thus long living connections (especially the masters) have been "killed" prematurely.

BTW: Now we monitor the number of states ;)

Sorry for the hassle. I jumped prematurely onto this ticket.

Cheers
